include/sycl/algorithm/reduce.hpp

```diff
@@ -94,6 +95,76 @@ typename std::iterator_traits<Iterator>::value_type reduce(
                                      cl::sycl::access::target::host_buffer>();
   return hI[0] + init;
 }
+#else
+template <class ExecutionPolicy, class Iterator, class T, class BinaryOperation>
+typename std::iterator_traits<Iterator>::value_type reduce(
+    ExecutionPolicy &sep, Iterator b, Iterator e, T init, BinaryOperation bop) {
+
+	typedef typename std::iterator_traits<Iterator>::value_type type_;
+
+	cl::sycl::queue q(sep.get_queue());
+	
+	auto vectorSize = sycl::helpers::distance(b, e);
+	
+	if (vectorSize < 1) {
+		return init;
+	}
+
+	auto device = q.get_device();
+	auto nb_work_item   = device.get_info<cl::sycl::info::device::max_work_group_size>();
+
+	auto buff = sycl::helpers::make_const_buffer(b, e);
+	
+	size_t length = vectorSize;
+	while (length > 1) {
+		q.submit([&] (cl::sycl::handler &cgh) {
+			cl::sycl::nd_range<1> rng { cl::sycl::range<1>{std::max(length, nb_work_item)},
+										cl::sycl::range<1>{nb_work_item}};
+
+			auto buff_acc = buff.template get_access<cl::sycl::access::mode::read_write>(cgh);
+
+			cl::sycl::accessor<type_, 1, cl::sycl::access::mode::read_write,
+			                             cl::sycl::access::target::local> scratch(cl::sycl::range<1>(nb_work_item), cgh);
+			cgh.parallel_for_work_group<typename ExecutionPolicy::kernelName>(rng, [=](cl::sycl::group<1> grp) {
+				
+				// Step 1:
+				//     copy a sub-vector of 'buff' into 'scratch'
+				grp.parallel_for_work_item([&](cl::sycl::nd_item<1> id) {
+					int globalid = id.get_global(0);
+					int localid = id.get_local(0);
+					if (globalid < length) {
+						scratch[localid] = buff_acc[globalid];
+					}
+				});	
+				
+				// Step 2:
+				//     reduce 'scratch' with the associative, commutative, binary operator 'bop'
+				//     use a parallel divide and conquer strategy
+				int min = (length < nb_work_item) ? length : nb_work_item;
+				for (int offset = min >> 1; offset > 0; offset = offset >> 1) {
+					grp.parallel_for_work_item([&](cl::sycl::nd_item<1> id){
+						int localid = id.get_local(0);
+						if (localid < offset) {
+							scratch[localid] = bop(scratch[localid], scratch[localid + offset]);
+						}
+					});
+				}
+
+				// Step 3:
+				//     copy back the reduction of 'scratch' into 'buff'
+				buff_acc[grp.get(0)] = scratch[0];
```

Comment:
After discussion together, it looks like there was already this race condition in the original algorithm you translated... So I guess it worths to change the algorithm here, while moving to hierarchical parallelism.